PUBDEV-8145: Fix R package based on CRAN feedback

ledell opened the pull request:
https://h2oai.atlassian.net/browse/PUBDEV-8145

tomasfryda reviewed (approved):
LGTM! Thank you @ledell!

michalkurka reviewed (changes requested):
download.file doesn't take argument 'timeout'

verified on `R version 3.6.3`

michalkurka commented on h2o-r/h2o-package/R/connection.R:
```diff
@@ -841,7 +838,8 @@ h2o.resume <- function(recovery_dir=NULL) {
   cat("(This could take a few minutes, please be patient...)\n")
-  download.file(url = h2o_url, destfile = temp_file, mode = "wb", cacheOK = FALSE, quiet = TRUE)
+  flush.console()
+  download.file(url = h2o_url, destfile = temp_file, mode = "wb", cacheOK = FALSE, quiet = TRUE, timeout = max(1000, getOption("timeout")))
```
timeout is in seconds, do we really want the user to wait up to 1000 seconds for everything to fail?

ledell replied:
@michalkurka That was the exact timeout duration that CRAN suggested, based on the size of the h2o.jar, so I think that's probably what we should leave it as.

ledell commented:
Here's what I changed: 
- Set timeout as an option in `.onLoad()`:  `options(timeout = max(1000, getOption("timeout")))`
- Consolidated the two copies of `.onLoad()` (only one was being used, this was a bug).
- Moved the `.onLoad()` into  `zzz.R`, which is the standard location for these functions.  I left `.onAttach()` et al in `connection.R` because there were some variables which were being used in both `.onAttach()` and the other connection.R internal functions and it seemed slightly risky to move.

tomasfryda commented on h2o-r/h2o-package/R/zzz.R:
```diff
@@ -1 +1,61 @@
+.onLoad <- function(lib, pkg) {
+  
+  # for file.download, used to download h2o.jar
+  options(timeout = max(1000, getOption("timeout")))
```
I'd suggest to use the same pattern as is often used for `par`:

```r
old_options <- options(timeout = max(1000, getOption("timeout")))  # NOTE: old_options is a named list containing only the old values of the changed options
on.exit(options(old_options))
```

This way the timeout is set only within this function scope so we'd need to move the `.h2o.downloadJar()` inside this function. 

I think it shouldn't cause `.h2o.downloadJar()` to be ran more times than it used to but I am not sure; my assumptions are:
>Now it is called on every `source` which would be just once per session and this would change to be called just once per load which should also be just once per session (both statements should be valid unless a user decides to unload the pkg and reload it which should be relatively rare).

jangorecki replied:
I think it is better to move setting option to `.h2o.downloadJar`

ledell replied:
@tomasfryda Are you saying that you want to change things around so that the timeout only applies to the `.h2o.downloadJar()` function?  

@jangorecki Why do you suggest that we move the option setting to `.h2o.downloadJar`?  The same reason above (so it only applies to one function)?

ledell replied:
Here's the current behavior:

```
> library(h2o)

----------------------------------------------------------------------

Your next step is to start H2O:
    > h2o.init()

For H2O package documentation, ask for help:
    > ??h2o

After starting H2O, you can use the Web UI at http://localhost:54321
For more information visit https://docs.h2o.ai

----------------------------------------------------------------------


Attaching package: ‘h2o’

The following objects are masked from ‘package:stats’:

    cor, sd, var

The following objects are masked from ‘package:base’:

    &&, %*%, %in%, ||, apply, as.factor, as.numeric, colnames,
    colnames<-, ifelse, is.character, is.factor, is.numeric, log,
    log10, log1p, log2, round, signif, trunc

> getOption("timeout")
[1] 1000
> detach("package:h2o", unload=TRUE)
[1] "A shutdown has been triggered. "
> getOption("timeout")
[1] 1000
```
It will go back to 60 seconds only with a new R session.

jangorecki replied:
Yes, so current behavior can be improved

tomasfryda replied:
@ledell yes, I would limit side effects like this to the smallest possible extent so we don't affect anything else, e.g., some other library the same session.

michalkurka commented:
@ledell to pass the "Check Pull Request Metadata" please change any Java file (eg. docuchange, anything)

currently there is a bug that doesn't let you pass the check if you only make a change in R files

ledell commented on h2o-r/h2o-package/R/connection.R:
```diff
@@ -758,7 +742,10 @@ h2o.resume <- function(recovery_dir=NULL) {
+#' @importFrom utils flush.console
 .h2o.downloadJar <- function(overwrite = FALSE) {
+  old_options <- options(timeout = max(1000, getOption("timeout")))
+  on.exit(options(old_options))
```
@tomasfryda  Can you explain what this does?  To me, it looks like it will save the "new" timeout option of 1000s (but you call it `old_options`).  And then it sets the timeout to 1000s only when the `.h2o.downloadJar()` exits (which would be after the download instead of before it).  It seems like it does the opposite of what we want it to do?  I might be missing something though.

Would we do something like this instead?
```
old_options <- options(timeout = getOption("timeout")) #save what the user had
options(timeout = max(1000, getOption("timeout"))) #set the 1000s for the download
on.exit(options(old_options)) #return to the previous settings
  ```

ledell replied:
cc @jangorecki Can you confirm?

tomasfryda replied:
@ledell From the options documentation(https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/options)

>For options(), a list of all set options sorted by name. For options(name), a list of length one containing the set value, or NULL if it is unset. **For uses setting one or more options, a list with the previous values of the options changed (returned invisibly).**

ledell replied:
@tomasfryda Ok, that's not what I expected!  Thanks for the reference.  

We can ignore my comments above -- the code does what we want it to.  So I think we are good to merge!

ledell commented:
@michalkurka I edited Java file, per your request.  Changing the Java file re-ran all the tests, but they were previously green, so I merged before the tests re-ran: http://mr-0xg1:8080/blue/organizations/jenkins/h2o-3-pipeline/detail/PR-5492/3/pipeline
